File: 표/1_dfs_bfs/네트워크.py
"""
n: 컴퓨터 개수 (1 <= n <= 200)
computers: 연결 정보를 담은 2차원 배열 (0 ~ n-1로 표현, 연결: 1)
- [[1, 1, 0], [1, 1, 0], [0, 0, 1]]에서 
  1,1,0은 각각 1,2,3번째 컴퓨터에 연결,연결,비연결 상태임을 의미
- computers[i][i]는 항상 1 -> 즉 자기 자신과는 항상 연결
- 아무것도 연결되지 않은 컴퓨터도 하나의 네트워크로 간주
Q. 네트워크의 개수는? 
"""

def solution(n, computers):
    answer = 0
    visited = [False]*n
    
    def dfs(i):
        nonlocal visited
        for j, connected in enumerate(computers[i]):
            if connected and not visited[j]:
                visited[j] = True
                dfs(j)
        
    for k in range(n):
        if not visited[k]:
            dfs(k)
            answer += 1
    
    return answer

File: 최윤경/네트워크.js
/*
[접근]

1. 연결된 컴퓨터들을 하나의 네트워크로 판단 → DFS/BFS
2. visited로 미방문 컴퓨터 확인 → answer + 1
3. DFS로 연결된 컴퓨터들을 모두 방문 처리
4. 최종 answer 반환
*/

function solution(n, computers) {
  let answer = 0;
  let visited = Array(n).fill(false);

  function dfs(computer) {
    visited[computer] = true;

    for (let i = 0; i < n; i++) {
      if (computers[computer][i] === 1 && !visited[i]) {
        dfs(i);
      }
    }
  }

  for (let i = 0; i < n; i++) {
    if (!visited[i]) {
      answer++;
      dfs(i);
    }
  }

  return answer;
}
